### PR TITLE
Fouque Tibouchi Hash to Curve

### PR DESCRIPTION
**Pull request by kilic**
This pr is an attemption to 

- [x] remove unused bls functions
- [x] migrate to Fouque Tibouchi hash to curve
- [x] use [faster modexp](https://github.com/ChihChengLiang/modexp) functions for sqrt and inversion

### CODE REVIEW

**Comment by kilic** [comment]
Signature check for 32 tx.
operation gas cost: 5072117
transaction gas cost: 6820242

File: contracts/libs/BLS.sol [hunk 1]
```diff
@@ -95,34 +105,94 @@ library BLS {
...
+        bool success;
+        // solium-disable-next-line security/no-inline-assembly
+        assembly {
+            success := staticcall(sub(gas(), 2000), 6, bnAddInput, 128, p0, 64)
+            switch success
+                case 0 {
+                    invalid()
+                }
+        }
+        require(success, "");
```
Comment (ChihChengLiang) [inline contracts/libs/BLS.sol:130]: Should give reason here for better debugging.
```suggestion
        require(success, "Hash to point fail");
```

File: contracts/libs/BLS.sol [hunk 2]
```diff
@@ -95,34 +105,94 @@ library BLS {
...
-            y = mulmod(y, x, N);
-            y = addmod(y, 3, N);
-            (y, found) = sqrt(y);
-            if (found) {
-                p[0] = x;
-                p[1] = y;
-                break;
+        require(_x < N, "mapToPointFT: invalid field element");
+        uint256 x = _x;
+        bool decision = isNonResidueFP(x);
```
Comment (ChihChengLiang) [inline contracts/libs/BLS.sol:141]: It looks like we can do isNonResidueFP in modexp optimization too

File: contracts/libs/BLS.sol [hunk 3]
```diff
@@ -410,24 +324,28 @@ library BLS {
         return !isNonResidue;
     }
 
-    function isNonResidueFP2(uint256[2] memory e)
+    function sqrtFaster(uint256 xx)
         internal
         view
```
Comment (ChihChengLiang) [inline contracts/libs/BLS.sol:329]: This function is pure
```suggestion
        pure
```

File: contracts/libs/BLS.sol [hunk 4]
```diff
@@ -442,25 +360,29 @@ library BLS {
...
+            x := mload(freemem)
+            hasRoot := eq(xx, mulmod(x, x, N))
         }
-        require(callSuccess, "BLS: isNonResidueFP2 modexp call failed");
-        return !isNonResidue;
+        require(callSuccess, "BLS: sqrt modexp call failed");
     }
 
-    function sqrt(uint256 xx) internal view returns (uint256 x, bool hasRoot) {
+    function inverseFaster(uint256 a) internal view returns (uint256) {
```
Comment (ChihChengLiang) [inline contracts/libs/BLS.sol:369]: This function is pure too
```suggestion
    function inverseFaster(uint256 a) internal pure returns (uint256) {
```

File: ts/hash_to_field.ts [hunk 5]
```diff
@@ -0,0 +1,209 @@
...
+    const msg = "0x616263";
+    mcl.setDomain(dst);
+    const p = mcl.hashToPoint(msg);
+    const result = mcl.g1ToHex(p);
+    assert.equal(result[0], expected0);
+    assert.equal(result[1], expected1);
+}
+
+import * as mcl from "./mcl";
+async function test() {
```
Comment (ChihChengLiang) [inline ts/hash_to_field.ts:201]: Can we organize these test related functions to the `test/`?

**Review by ChihChengLiang — COMMENTED** [review]
My first pass of review

File: ts/hash_to_field.ts (hunk 5 above)
Comment (kilic) [reply ts/hash_to_field.ts:201]: those tests test tests :). 
I've left them for reviewers or auditors to prove that we are in line with the standart. These test don't need be run repeatedly.

Comment (ChihChengLiang) [reply ts/hash_to_field.ts:201]: I see. I'll later organize them to #202.

**Review by ChihChengLiang — APPROVED** [review]
Looks good to me

**Pull request merged**